src/core/assert/dev.js

```diff
@@ -0,0 +1,209 @@
+/**
+ * Copyright 2021 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as assertions from './base';
+import {isMinifiedMode} from '../minified-mode';
+
+/**
+ * @fileoverview This file provides the entrypoint for dev assertions. It's
+ * designed so all functions are pure function calls to improve inlining.
+ */
+
+/**
+ * Throws an error if the first argument isn't trueish. Mirrors devAssert in
+ * src/log.js.
+ * @param {T} shouldBeTruthy
+ * @param {string=} opt_message
+ * @param {*=} opt_1 Optional argument (var arg as individual params for better
+ * @param {*=} opt_2 Optional argument inlining)
+ * @param {*=} opt_3 Optional argument
+ * @param {*=} opt_4 Optional argument
+ * @param {*=} opt_5 Optional argument
+ * @param {*=} opt_6 Optional argument
+ * @param {*=} opt_7 Optional argument
+ * @param {*=} opt_8 Optional argument
+ * @param {*=} opt_9 Optional argument
+ * @return {T}
+ * @template T
+ * @throws {Error} when shouldBeTruthy is not truthy.
+ * @closurePrimitive {asserts.truthy}
+ */
+export function assert(
+  shouldBeTruthy,
+  opt_message,
+  opt_1,
+  opt_2,
+  opt_3,
+  opt_4,
+  opt_5,
+  opt_6,
+  opt_7,
+  opt_8,
+  opt_9
+) {
+  if (isMinifiedMode()) {
+    return shouldBeTruthy;
+  }
+
+  if (self.__AMP_ASSERTION_CHECK) {
+    // This will never execute regardless, but will be included on unminified
+    // builds. It will be DCE'd away from minified builds, and so can be used to
+    // validate that Babel is properly removing dev assertions in minified
+    // builds.
+    console /*OK*/
+      .log('__devAssert_sentinel__');
+  }
+
+  return assertions.assert(
+    '',
+    shouldBeTruthy,
+    opt_message,
+    opt_1,
+    opt_2,
+    opt_3,
+    opt_4,
+    opt_5,
+    opt_6,
+    opt_7,
+    opt_8,
+    opt_9
+  );
+}
+
+/**
+ * Throws an error if the first argument isn't an Element.
+ *
+ * For more details see `assert`.
+ *
+ * @param {?string} sentinel
+ * @param {*} shouldBeElement
+ * @param {!Array<*>|string=} opt_message The assertion message
+ * @return {!Element} The value of shouldBeTrueish.
+ * @throws {Error} when shouldBeElement is not an Element
+ * @closurePrimitive {asserts.matchesReturn}
+ */
+export function assertElement(sentinel, shouldBeElement, opt_message) {
+  return assertions.assertElement(
+    /** @type {!AssertionFunction} */ (assert),
```

Comment:
I'm concerned that none of these are going to be properly DCE'd. The indirection here requires Closure to understand that `assert` is empty, so `assertElement`'s call to `assert` is empty, so `assertElement` is empty, which is difficult (and likely impossible to do if we switch to Terser).

Instead, if the intention is for `dev.*` to never throw, explicitly check for `isMinifiedMode` is each.

Reply:
Looked at `dist` output and you were right. Dropped the `ifMinifiedMode` bailout into all the dev helpers; looks like they're DCEing correctly based on some manual tests (they aren't used anywhere yet so they won't show up in bundle-size bot or v0.js output)